src/query/functions/temporal/base_test.go

```diff
@@ -351,6 +353,158 @@ func setup(numBlocks int, duration time.Duration, nextBound int) *testContext {
+
+	for _, mockBuilder := range builderCtx.MockBlockBuilders {
+		assert.True(t, mockBuilder.BuiltBlock.Closed)
+	}
+}
+
+func TestProcessCompletedBlocks_ClosesBlocksOnError(t *testing.T) {
+	numBlocks := 2
+	tc := setup(numBlocks, 5*time.Minute, 1)
+	ctrl := gomock.NewController(t)
+	setupCloseableBlock(ctrl, tc.Node)
+
+	testErr := errors.New("test err")
+	tc.Blocks[1] = blockWithDownstreamErr{Block: tc.Blocks[1], Err: testErr}
+
+	processRequests := make([]processRequest, numBlocks)
+	for i, blk := range tc.Blocks {
+		unconsolidated, err := blk.Unconsolidated()
+		require.NoError(t, err)
+
+		processRequests[i] = processRequest{
+			queryCtx: models.NoopQueryContext(),
+			blk:      unconsolidated,
+			bounds:   tc.Bounds,
+			deps:     nil,
+		}
+	}
+
+	blocks, err := tc.Node.processCompletedBlocks(models.NoopQueryContext(), processRequests, numBlocks)
+	require.EqualError(t, err, testErr.Error())
+
+	for _, bl := range blocks {
+		assert.True(t, bl.(*closeSpyBlock).Closed)
+	}
+}
+
+type closeableBlockBuilderContext struct {
+	MockController    *Mockcontroller
+	MockBlockBuilders []*closeSpyBlockBuilder
+}
+
+// setupCloseableBlock mocks out node.controller to return a block builder which
+// builds closeSpyBlock instances, so that you can inspect whether
+// or not a block was closed (using closeSpyBlock.Closed). See TestBaseClosesBlocks
+// for an example.
+func setupCloseableBlock(ctrl *gomock.Controller, node *baseNode) closeableBlockBuilderContext {
+	mockController := NewMockcontroller(ctrl)
+	mockBuilders := make([]*closeSpyBlockBuilder, 0)
+
+	mockController.EXPECT().Process(gomock.Any(), gomock.Any()).Return(nil)
+
+	// return a regular ColumnBlockBuilder, wrapped with closeSpyBlockBuilder
+	mockController.EXPECT().BlockBuilder(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(
+			queryCtx *models.QueryContext,
+			blockMeta block.Metadata,
+			seriesMeta []block.SeriesMeta) (block.Builder, error) {
+			mb := &closeSpyBlockBuilder{
+				Builder: block.NewColumnBlockBuilder(blockMeta, seriesMeta),
+			}
+			mockBuilders = append(mockBuilders, mb)
+			return mb, nil
+		})
+
+	node.controller = mockController
+
+	return closeableBlockBuilderContext{
+		MockController:    mockController,
+		MockBlockBuilders: mockBuilders,
+	}
+}
+
+// closeSpyBlockBuilder wraps a block.Builder to build a closeSpyBlock
+// instead of a regular block. It is otherwise equivalent to the wrapped Builder.
+type closeSpyBlockBuilder struct {
+	block.Builder
+
+	BuiltBlock *closeSpyBlock
+}
+
+func (bbc *closeSpyBlockBuilder) Build() block.Block {
+	bbc.BuiltBlock = &closeSpyBlock{
+		Block: bbc.Builder.Build(),
+	}
+	return bbc.BuiltBlock
+}
+
+// closeSpyBlock wraps a block.Block to allow assertions on the Close()
+// method.
+type closeSpyBlock struct {
+	block.Block
+
+	Closed bool
+}
+
+func (bwc *closeSpyBlock) Close() error {
+	bwc.Closed = true
```

Comment:
If you're doing this anyway, might as well make `closed` an int and append it to make sure we close exactly once